### ISSUE TITLE
Is this ready for experimental use?

### ISSUE THREAD

**Issue by ebigram** [hbutani/spark-sql-macros#1]
I went through the docs and some of the source, and that was a fascinating edifying read; obviously having first-class UDFs would open up all kinds of new development opportunities with Spark, so this is potentially very exciting(!)

I'd love to give it a go for am ancillary job to see for myself, but don't quite have the time to really delve into the gory details of a brand new library. I didn't see much setup instructions in the wiki, and I'd be glad to be corrected, I was just wondering if I should hold off as you're still adding some documentation or if it should be pretty straightforward to work with in its current state. 

Would love to loop back to this and help with said documentation/debugging, just not right at this moment. Cool stuff nonetheless, and I'll be sure to follow it.

**Comment by ebigram** [comment]
Somehow missed the pre-release 🤦, so Spark 3.0 and adding the jar dependency might be enough to give it an honest shake.

**Comment by hbutani** [comment]
Yes adding the jar to your Spark env should be enough.
Have been developing with Spark 3.1 and scala 2.12; but I think it should work with Spark 3.0 and maybe earlier releases.

Nothing special in terms of setup; start with  `import org.apache.spark.sql.defineMacros._` which will
add the implicit functions on SparkSession to register and define macros.

**Comment by ebigram** [comment]
FWIW, I'm using Scala 2.13 and Spark 2.4.0,, importing `defineMacros._` like you said  and copying your example verbatim:

```scala
import org.apache.spark.sql.defineMacros._

spark.registerMacro("intUDM", spark.udm((i: Int) => {
   val j = 2
   i + j
  }))
```

I get this error:

````scala
Error: exception during macro expansion: java.lang.ClassNotFoundException: org.apache.spark.sql.catalyst.expressions.objects.UnresolvedCatalystToExternalMap at java.net.URLClassLoader.findClass(URLClassLoader.java:382) at java.lang.ClassLoader.loadClass(ClassLoader.java:418) at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352) at java.lang.ClassLoader.loadClass(ClassLoader.java:351) at org.apache.spark.sql.sqlmacros.MacrosEnv.$init$(MacrosEnv.scala:30) at org.apache.spark.sql.sqlmacros.SQLMacro.<init>(SQLMacro.scala:26) (Line 3)
````

I'm running this in the repl which may be complicating things. I will try to bump up my versions. Just in case it rings a bell.

**Comment by hbutani** [comment]
`org.apache.spark.sql.catalyst.expressions.objects.UnresolvedCatalystToExternalMap ` was introduced in SPARK-25817 which resolved for 3.0. 

Based on your attempt it is clear that you need Spark 3.0 at least.
I have been developing with Spark 3.1, so if you can start at Spark 3.1.0.0

**Comment by ebigram** [comment]
Ok thanks, closing this for now as it's probably some issue on my end.
